clients/drcachesim/tracer/instr_counter.cpp

```diff
@@ -347,9 +398,122 @@ event_inscount_thread_init(void *drcontext)
+    }
+
+    // Enforcing constraints on intervals:
+    // 1) they need to be ordered by start time.
+    std::sort(instr_intervals.begin(), instr_intervals.end(), cmp_start_instr);
+
+    // 2) overlapping intervals must be merged.
+    std::vector<instr_interval_t> instr_intervals_merged;
+    instr_intervals_merged.emplace_back(instr_intervals[0]);
+    for (instr_interval_t &interval : instr_intervals) {
+        uint64 end = interval.start + interval.duration;
+        instr_interval_t &last_interval = instr_intervals_merged.back();
+        uint64 last_end = last_interval.start + last_interval.duration;
+        if (interval.start <= last_end) {
+            uint64 max_end = std::max(last_end, end);
+            last_interval.duration = max_end - last_interval.start;
+        } else {
+            instr_intervals_merged.emplace_back(interval);
+        }
+    }
+
+    return instr_intervals_merged;
+}
+
+// Transforms instruction intervals from <start,duration> pairs to trace and no_trace
+// number of instructions. Has the side effect of populating the read-only, global vectors
+// trace_for_instrs_list and no_trace_for_instrs_list.
+static void
+compute_trace_intervals(std::vector<instr_interval_t> &instr_intervals)
+{
+    if (instr_intervals.empty())
+        return;
+
+    trace_for_instrs_list.resize(instr_intervals.size() + 1, 0);
+    trace_for_instrs_list[0] = instr_intervals[0].duration;
+    no_trace_for_instrs_list.resize(instr_intervals.size() + 1, DELAY_FOREVER_THRESHOLD);
+    no_trace_for_instrs_list[0] = instr_intervals[0].start;
+    for (uint i = 1; i < instr_intervals.size(); ++i) {
+        uint64 trace_for_instrs = instr_intervals[i].duration;
+        trace_for_instrs_list[i] = trace_for_instrs;
+        uint64 no_trace_for_instrs = instr_intervals[i].start -
+            (instr_intervals[i - 1].start + instr_intervals[i - 1].duration);
+        no_trace_for_instrs_list[i] = no_trace_for_instrs;
+    }
+}
+
 void
 event_inscount_init()
 {
+    std::string path_to_file = op_trace_instr_intervals_file.get_value();
+    if (!path_to_file.empty()) {
+        // Other instruction interval options (i.e., -trace_after_instrs,
+        // -trace_for_instrs, -retrace_every_instrs) are not compatible with
+        // -trace_instr_intervals_file. Check that they are not set.
+        if (op_trace_after_instrs.get_value() > 0 ||
+            op_trace_for_instrs.get_value() > 0 ||
+            op_retrace_every_instrs.get_value() > 0) {
+            FATAL("Fatal error: -trace_instr_intervals_file cannot be used with "
+                  "-trace_after_instrs, -trace_for_instrs, or -retrace_every_instrs.\n");
+        }
+        // Parse intervals file.
+        std::vector<instr_interval_t> instr_intervals =
+            parse_instr_intervals_file(op_trace_instr_intervals_file.get_value());
+        // Populate the [no_]trace_for_instrs_list global vectors.
+        compute_trace_intervals(instr_intervals);
```

Comment:
Would it make sense to merge the handling of trace_after_instrs and trace_for_instrs into trace_for_instrs_list and no_trace_for_instrs_list?

Reply:
We could, although it might make the code less maintainable/more confusing?
Basically we'd have:
```
no_trace_for_instrs_list[0] = op_trace_after_instrs.get_value();
trace_for_instrs_list[0] = op_trace_for_instrs.get_value();
no_trace_for_instrs_list[1] = op_retrace_every_instrs.get_value();
trace_for_instrs_list[1] = op_trace_for_instrs.get_value();
```
and then we would increment `instr_intervals_idx` only after the first window.

Reply:
> and then we would increment instr_intervals_idx only after the first window.

Even with the new option we'd have to ensure that we don't increment instr_intervals_idx beyond what's specified, right? So that same logic would also work for a size of 1?

Reply:
This is actually more complicated than I anticipated.
I implemented it here https://github.com/DynamoRIO/dynamorio/pull/6937 if you want to take a look, but it complicated the whole logic quite a bit.
Not worth it in my opinion.

> So that same logic would also work for a size of 1?

One of the issues is that it's not really size == 1 (single window). We still have multiple, periodic windows.